kasa/discover.py

```diff
@@ -315,5 +312,64 @@ def _get_device_class(info: dict) -> Type[SmartDevice]:
                 return SmartLightStrip
 
             return SmartBulb
+        raise UnsupportedDeviceException("Unknown device type: %s" % type_)
+
+    @staticmethod
+    def _get_device_instance_legacy(data: bytes, ip: str, port: int) -> SmartDevice:
+        """Get SmartDevice for device described by passed data from legacy 9999 response."""
+        try:
+            info = json_loads(TPLinkSmartHomeProtocol.decrypt(data))
+        except Exception as ex:
+            raise SmartDeviceException(
+                f"Unable to read response from device: {ip}: {ex}"
+            )
+
+        _LOGGER.debug("[DISCOVERY] %s << %s", ip, info)
+
+        device_class = Discover._get_device_class(info)
+        device = device_class(ip, port=port)
+        device.update_from_discover_info(info)
+        return device
 
-        raise SmartDeviceException("Unknown device type: %s" % type_)
+    @staticmethod
+    def _get_device_instance(
+        data: bytes, ip: str, port: int, credentials: Credentials
+    ) -> SmartDevice:
+        """Get SmartDevice for device described by passed data from new 20002 response."""
+        try:
+            info = json_loads(data[16:])
+        except Exception as ex:
+            raise SmartDeviceException(
+                f"Unable to read response from device: {ip}: {ex}"
+            )
+
+        if (
+            "result" in info
+            and "mgt_encrypt_schm" in info["result"]
+            and "encrypt_type" in info["result"]["mgt_encrypt_schm"]
+            and info["result"]["mgt_encrypt_schm"]["encrypt_type"] == "KLAP"
+            and "lv" not in info["result"]["mgt_encrypt_schm"]
+            and "device_type" in info["result"]
+        ):
+            type_ = info["result"]["device_type"]
+            device_class = None
+            if type_.upper() == "IOT.SMARTPLUGSWITCH":
+                device_class = SmartPlug
+
+            if device_class:
+                _LOGGER.debug("[DISCOVERY] %s << %s", ip, info)
+                device = device_class(ip, port=port, credentials=credentials)
+                # Set the MAC so HA can add the device and try authentication later
+                device._requires_update_overrides = {
+                    "mac": info["result"].get("mac"),
+                    "alias": ip,
+                    "model": info["result"].get("device_model"),
+                }
```

Comment:
See https://github.com/python-kasa/python-kasa/pull/477#discussion_r1324994623 – would it make sense to handle this inside the device class? We can adjust the update_from_sysinfo if needed, as that's something we need to have anyway in case tapo support gets added at some later point.

Reply:
How do you see that working because at the moment update_from_sysinfo just sets the whole _sys_info variable but in this case (and for HA really) we only have a few variables and they're in a different schema to regular sys_info?